src/System.Data.SqlClient/src/Microsoft/SqlServer/Server/SmiMetaData.cs

```diff
@@ -0,0 +1,854 @@
+                "nchar",                // SqlDbType.NChar
+                "ntext",                // SqlDbType.NText
+                "nvarchar",             // SqlDbType.NVarChar
+                "real",                 // SqlDbType.Real
+                "uniqueidentifier",     // SqlDbType.UniqueIdentifier
+                "smalldatetime",        // SqlDbType.SmallDateTime
+                "smallint",             // SqlDbType.SmallInt
+                "smallmoney",           // SqlDbType.SmallMoney
+                "text",                 // SqlDbType.Text
+                "timestamp",            // SqlDbType.Timestamp
+                "tinyint",              // SqlDbType.TinyInt
+                "varbinary",            // SqlDbType.VarBinary
+                "varchar",              // SqlDbType.VarChar
+                "sql_variant",          // SqlDbType.Variant
+                null,                   // placeholder for 24
+                "xml",                  // SqlDbType.Xml
+                null,                   // placeholder for 26
+                null,                   // placeholder for 27
+                null,                   // placeholder for 28
+                String.Empty,           // SqlDbType.Udt  -- get type name from Type.FullName instead.
+                String.Empty,           // Structured types have user-defined type names.
+                "date",                 // SqlDbType.Date
+                "time",                 // SqlDbType.Time
+                "datetime2",            // SqlDbType.DateTime2
+                "datetimeoffset",       // SqlDbType.DateTimeOffset
+            };
+
+        // Internal setter to be used by constructors only!  Modifies state!
+        private void SetDefaultsForType(SqlDbType dbType)
+        {
+            SmiMetaData smdDflt = GetDefaultForType(dbType);
+            _databaseType = dbType;
+            _maxLength = smdDflt.MaxLength;
+            _precision = smdDflt.Precision;
+            _scale = smdDflt.Scale;
+            _localeId = smdDflt.LocaleId;
+            _compareOptions = smdDflt.CompareOptions;
+            _isMultiValued = smdDflt._isMultiValued;
+            _fieldMetaData = smdDflt._fieldMetaData;            // This is ok due to immutability
+            _extendedProperties = smdDflt._extendedProperties;  // This is ok due to immutability
+        }
+    }
+
+
+    // SmiExtendedMetaData
+    //
+    //  Adds server-specific type extension information to base metadata, but still portable across a specific server.
+    //
+    internal class SmiExtendedMetaData : SmiMetaData
+    {
+        private string _name;           // context-dependant identifier, ie. parameter name for parameters, column name for columns, etc.
+
+        // three-part name for typed xml schema and for udt names
+        private string _typeSpecificNamePart1;
+        private string _typeSpecificNamePart2;
+        private string _typeSpecificNamePart3;
```

Comment:
Looks like most/all of the fields in this type hierarchy could be made readonly.  I'll stop commenting on them.